modules/500-openvpn/images/openvpn/entrypoint/entrypoint.go

```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2023 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"golang.org/x/sys/unix"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/vishvananda/netlink"
+)
+
+type iptablesRule struct {
+	Table string
+	Chain string
+	Rule  []string
+	Pos   int
+}
+
+func main() {
+	network, ok := os.LookupEnv("TUNNEL_NETWORK")
+	if !ok {
+		log.Fatal("The TUNNEL_NETWORK environment variable does not exist.")
+	}
+
+	protocol, ok := os.LookupEnv("OPENVPN_PROTO")
+	if !ok {
+		log.Fatal("The OPENVPN_PROTO environment variable does not exist.")
+	}
+
+	var mgmtport string
+	var routeTable int
+
+	switch protocol {
+	case "tcp":
+		mgmtport = "8989"
+		routeTable = 10
+	case "udp":
+		mgmtport = "9090"
+		routeTable = 11
+	default:
+		log.Fatalf("OPENVPN_PROTO env value must be set to 'tcp' or 'udp': %s", protocol)
+	}
+
+	devTunName := fmt.Sprintf("tun-%s", protocol)
+
+	iptablesMgr, err := iptables.NewWithProtocol(iptables.ProtocolIPv4)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	nat := &iptablesRule{
+		Table: "nat",
+		Chain: "POSTROUTING",
+		Rule:  strings.Fields(fmt.Sprintf("-s %s ! -d %s -j MASQUERADE", network, network)),
+		Pos:   1,
+	}
+
+	manglePreroutingSetMark := &iptablesRule{
+		Table: "mangle",
+		Chain: "PREROUTING",
+		Rule:  strings.Fields(fmt.Sprintf("-i %s -j CONNMARK --set-mark %d", devTunName, routeTable)),
+		Pos:   1,
+	}
+
+	manglePreroutingRestoreMark := &iptablesRule{
+		Table: "mangle",
+		Chain: "PREROUTING",
+		Rule:  strings.Fields("! -i tun+ -j CONNMARK --restore-mark"),
+		Pos:   2,
+	}
+	mangleOutputRestoreMark := &iptablesRule{
+		Table: "mangle",
+		Chain: "OUTPUT",
+		Rule:  strings.Fields("-j CONNMARK --restore-mark"),
+		Pos:   1,
+	}
+
+	err = insertUnique(iptablesMgr, nat)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = insertUnique(iptablesMgr, manglePreroutingSetMark)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = insertUnique(iptablesMgr, manglePreroutingRestoreMark)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = insertUnique(iptablesMgr, mangleOutputRestoreMark)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	rule := netlink.NewRule()
+	rule.Table = routeTable
+	rule.Priority = 1
+	rule.Mark = routeTable
+	err = netlink.RuleAdd(rule)
+
+	if err != nil {
+		if !os.IsExist(err) {
+			log.Fatal(err)
+		}
```

Comment:
Maybe add error logging to else?

Reply:
Everything is logged here, except when the rule exists. If the rule exists, then in my opinion this is not an error.